File: src/services/wasteScanning.ts
import { GoogleGenerativeAI } from "@google/generative-ai";

// Initialize Gemini
const genAI = new GoogleGenerativeAI(process.env.NEXT_PUBLIC_GEMINI_API_KEY || '');

// Add error handling for the waste scanning service
export async function scanWaste(imageData: string) {
    try {
        // Convert base64 to Uint8Array for Gemini
        const imageBytes = base64ToBytes(imageData);

        // Initialize the model
        const model = genAI.getGenerativeModel({ model: "gemini-pro-vision" });

        // Create prompt for waste classification
        const prompt = `Analyze this image and provide:
            1. Waste Category: (Recyclable/Organic/Non-recyclable)
            2. Explanation: Why it belongs in this category
            3. Disposal Instructions: How to properly dispose of this item
            4. Environmental Impact: Brief impact if not disposed properly`;

        // Prepare the image part
        const imagePart = {
            inlineData: {
                data: imageBytes,
                mimeType: "image/jpeg"
            },
        };

        // Generate content
        const result = await model.generateContent([prompt, imagePart]);
        const response = await result.response;
        const text = response.text();

        return {
            classification: text,
            timestamp: new Date().toISOString()
        };

    } catch (error) {
        console.error('Gemini scanning error:', error);
        throw new Error('Failed to scan waste: ' + (error as Error).message);
    }
}

// Helper function to convert base64 to bytes
function base64ToBytes(base64: string): Uint8Array {
    // Remove data URL prefix if present
    const base64String = base64.split(',')[1] || base64;
    const binaryString = window.atob(base64String);
    const bytes = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
}

// Add retry logic
async function retryOperation(operation: () => Promise<any>, maxRetries = 3) {
    for (let i = 0; i < maxRetries; i++) {
        try {
            return await operation();
        } catch (error) {
            if (i === maxRetries - 1) throw error;
            // Wait for 2^i * 1000 ms before retrying (exponential backoff)
            await new Promise(resolve => setTimeout(resolve, Math.pow(2, i) * 1000));
            console.log(`Retry attempt ${i + 1} of ${maxRetries}`);
        }
    }
}

export function validateImage(file: File): boolean {
    // Check file type
    const validTypes = ['image/jpeg', 'image/png', 'image/jpg'];
    if (!validTypes.includes(file.type)) {
        throw new Error('Invalid file type. Please upload a JPEG or PNG image.');
    }

    // Check file size (5MB limit)
    const maxSize = 5 * 1024 * 1024; // 5MB in bytes
    if (file.size > maxSize) {
        throw new Error('File too large. Please upload an image smaller than 5MB.');
    }

    return true;
} 
} 

File: lib/services/waste_scanning.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:google_generative_ai/google_generative_ai.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';

class WasteScanningService {
  late final GenerativeModel model;
  
  WasteScanningService() {
    model = GenerativeModel(
      model: 'gemini-1.5-flash',
      apiKey: dotenv.env['GEMINI_API_KEY']!,
      generationConfig: GenerationConfig(
        temperature: 0.4,
        topK: 32,
        topP: 1,
        maxOutputTokens: 4096,
      ),
    );
  }

  Future<ScanResult> scanWaste(Uint8List imageBytes) async {
    try {
      const prompt = '''
        Act as a waste management expert. Analyze this image and provide a detailed classification:

        Format your response EXACTLY as follows:
        OBJECT: [Identify the main item]
        CATEGORY: [Classify as: Recyclable/Non-Recyclable/Toxic/Organic]
        DISPOSAL: [Step-by-step disposal instructions]
        IMPACT: [Environmental impact if not disposed properly]
        TIPS: [3 practical recycling or eco-friendly tips]

        Be specific and practical in your response.
      ''';

      final content = Content.multi([
        TextPart(prompt),
        DataPart('image/jpeg', imageBytes),
      ]);

      final response = await model.generateContent(content);
      final text = response.text ?? '';
      
      // Parse the structured response
      final Map<String, String> parsedResult = {};
      final lines = text.split('\n');
      for (var line in lines) {
        if (line.contains(':')) {
          final parts = line.split(':');
          final key = parts[0].trim();
          final value = parts.sublist(1).join(':').trim();
          parsedResult[key] = value;
        }
      }

      return ScanResult(
        object: parsedResult['OBJECT'] ?? 'Unknown object',
        category: parsedResult['CATEGORY'] ?? 'Unknown category',
        disposal: parsedResult['DISPOSAL'] ?? 'No disposal information',
        impact: parsedResult['IMPACT'] ?? 'No impact information',
        tips: parsedResult['TIPS'] ?? 'No tips available',
        timestamp: DateTime.now(),
      );
    } catch (error) {
      print('Gemini scanning error: $error');
      throw Exception('Failed to scan waste: $error');
    }
  }

  bool validateImage(Uint8List imageBytes) {
    // Check file size (5MB limit)
    const maxSize = 5 * 1024 * 1024; // 5MB in bytes
    if (imageBytes.length > maxSize) {
      throw Exception('File too large. Please upload an image smaller than 5MB.');
    }
    return true;
  }
}

class ScanResult {
  final String object;
  final String category;
  final String disposal;
  final String impact;
  final String tips;
  final DateTime timestamp;

  ScanResult({
    required this.object,
    required this.category,
    required this.disposal,
    required this.impact,
    required this.tips,
    required this.timestamp,
  });

  String get fullClassification => '''
    🗑️ Object: $object
    ♻️ Category: $category
    🚮 Disposal: $disposal
    🌍 Impact: $impact
    💡 Tips: $tips
  ''';
} 